processing/src/main/java/org/apache/druid/segment/join/JoinFilterAnalyzer.java

```diff
@@ -0,0 +1,743 @@
+        retainRhs,
+        orFilter,
+        new OrFilter(newFilters),
+        null
+    );
+  }
+
+  /**
+   * Rewrites a selector filter on a join table into an IN filter on the base table.
+   *
+   * @param baseAdapter      The adapter for the join
+   * @param selectorFilter   SelectorFilter to be rewritten
+   * @param prefixes         Map of join table prefixes to clauses
+   * @param equiconditions   Map of equiconditions
+   * @param correlationCache Cache of column correlation analyses
+   *
+   * @return A JoinFilterAnalysis that indicates how to handle the potentially rewritten filter
+   */
+  private static JoinFilterAnalysis rewriteSelectorFilter(
+      HashJoinSegmentStorageAdapter baseAdapter,
+      SelectorFilter selectorFilter,
+      Map<String, JoinableClause> prefixes,
+      Map<String, Expr> equiconditions,
+      Map<String, List<JoinFilterColumnCorrelationAnalysis>> correlationCache
+  )
+  {
+    String filteringColumn = selectorFilter.getDimension();
+    for (Map.Entry<String, JoinableClause> prefixAndClause : prefixes.entrySet()) {
+      if (filteringColumn.startsWith(prefixAndClause.getKey())) {
+        List<JoinFilterColumnCorrelationAnalysis> correlations = correlationCache.computeIfAbsent(
+            prefixAndClause.getKey(),
+            p -> findCorrelatedBaseTableColumns(
+                baseAdapter,
+                p,
+                prefixes.get(p),
+                equiconditions
+            )
+        );
+
+        if (correlations == null) {
+          return JoinFilterAnalysis.createNoPushdownFilterAnalysis(selectorFilter);
+        }
+
+        List<Filter> newFilters = new ArrayList<>();
+        List<VirtualColumn> pushdownVirtualColumns = new ArrayList<>();
+
+        for (JoinFilterColumnCorrelationAnalysis correlationAnalysis : correlations) {
+          if (correlationAnalysis.supportsPushDown()) {
+            List<String> correlatedValues = getCorrelatedValuesForPushDown(
+                selectorFilter.getDimension(),
+                selectorFilter.getValue(),
+                correlationAnalysis.getJoinColumn(),
+                prefixAndClause.getValue()
+            );
+
+            if (correlatedValues == null) {
+              return JoinFilterAnalysis.createNoPushdownFilterAnalysis(selectorFilter);
+            }
+
+            for (String correlatedBaseColumn : correlationAnalysis.getBaseColumns()) {
+              InFilter rewrittenFilter = (InFilter) new InDimFilter(
+                  correlatedBaseColumn,
+                  correlatedValues,
+                  null,
+                  null
+              ).toFilter();
+              newFilters.add(rewrittenFilter);
+            }
+
+            for (Expr correlatedBaseExpr : correlationAnalysis.getBaseExpressions()) {
+              // need to create a virtual column for the expressions when pushing down
+              String vcName = getCorrelatedBaseExprVirtualColumnName(pushdownVirtualColumns.size());
```

Comment:
I just realized there's a bug with this part (it needs to create a globally unique virtual column name), will fix in a follow-on PR.